icicle-compiler/src/Icicle/Compiler/Sea.hs

```diff
@@ -0,0 +1,192 @@
+
+  let
+    icolumn =
+      Runtime.InputColumn lens times (Storable.fromList tombstones) column
+
+    inputs =
+      Map.singleton iid icolumn
+
+  pure $
+    Runtime.Input entityKeys inputs
+
+fromOutputs :: Striped.Column -> Either CompilerSeaError [Common.BaseValue]
+fromOutputs column = do
+  xs <- bimap CompilerSeaStripedError Boxed.toList $ Striped.toLogical column
+  first CompilerSeaLogicalError $ traverse (Logical.toBaseValue (Schema.toValType (Striped.schema column))) xs
+
+fromFacts :: InputName -> Common.ValType -> [AsAt Fact] -> Either CompilerSeaError (Map Entity [AsAt Common.BaseValue])
+fromFacts name typ0 facts0 = do
+  let
+    facts1 =
+      List.filter ((== name) . factAttribute . atFact) facts0
+
+    typ1 =
+      Common.SumT Common.ErrorT typ0
+
+    fromFact (AsAt (Fact e _ x0) t) = do
+      x <- maybeToRight (CompilerSeaValueToCore x0 typ1) $ Common.valueToCore x0 typ1
+      pure (e, [AsAt x t])
+
+  facts2 <- traverse fromFact facts1
+  pure $ Map.fromListWith (flip (<>)) facts2
+
+seaEval ::
+     Common.EvalContext
+  -> [AsAt Fact]
+  -> Source.QueryTyped Source.Var
+  -> AvalProgramTyped  Source.Var Flat.Prim
+  -> EitherT CompilerSeaError IO [Result]
+seaEval ctx facts0 (renameQT unVar -> query) program = do
+  let
+    iid =
+      case Query.queryInput query of
+        QualifiedInput x ->
+          x
+        UnqualifiedInput name ->
+          InputId [namespace|default|] name
+
+    maxMapSize =
+      Runtime.MaximumMapSize . fromIntegral $ Common.evalMaxMapSize ctx
+
+    stime =
+      Runtime.SnapshotTime . Runtime.Time64 . packedOfTime $ Common.evalSnapshotTime ctx
+
+  context <- hoistEither . first CompilerSeaRuntimeError . Runtime.compileAvalanche $
+    Runtime.AvalancheContext "Icicle.Compiler.Sea.seaEval" (Map.singleton iid (program :| []))
+  runtime <- firstT CompilerSeaRuntimeError $ Runtime.compileSea Runtime.SkipJetskiCache context
+
+  cluster <- hoistEither . maybeToRight CompilerSeaNoInputs . listToMaybe . Map.elems $ Runtime.runtimeClusters runtime
+
+  let
+    inputType =
+      Schema.toValType . Runtime.clusterInputSchema $
+        Sea.clusterAnnotation cluster
+
+  facts1 <- hoistEither $ fromFacts (inputName iid) inputType facts0
+  input <- hoistEither $ fromInputs iid inputType facts1
+
+  output0 <- firstT CompilerSeaRuntimeError $ Runtime.snapshotBlock runtime maxMapSize stime input
+  output1 <- hoistEither . maybeToRight CompilerSeaNoOutputs . listToMaybe . Map.elems $ Runtime.outputColumns output0
+
+  let
+    entities =
+      fmap (Entity . Text.decodeUtf8 . Runtime.unEntityId . Runtime.entityId . Runtime.snapshotEntity) .
+      Boxed.toList $
+      Runtime.outputKey output0
+
+  output2 <- hoistEither $ fromOutputs output1
+  output3 <- hoistEither $ traverse (\x -> maybeToRight (CompilerSeaValueFromCore x) $ Common.valueFromCore x) output2
+
+  pure .
+    fmap Result $
+    List.zipWith ((,)) entities output3
```

Comment:
It's supposed to have the same type more or less as `coreEval` and `avalancheEval` in `Icicle.Compiler`, that's why I put it here fwiw. I just couldn't put it in `Icicle.Compiler` because of a circular dependency.